core-api/src/main/java/com/optimizely/ab/internal/DefaultLRUCache.java

```diff
@@ -0,0 +1,147 @@
+/**
+ *
+ *    Copyright 2022, Optimizely
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+package com.optimizely.ab.internal;
+
+import com.optimizely.ab.annotations.VisibleForTesting;
+
+import java.util.*;
+
+public class DefaultLRUCache<T> implements LRUCache<T> {
+
+    private final Object lock = new Object();
+
+    private Integer maxSize;
+
+    private Long timeoutMillis;
+
+    final LinkedList<ItemWrapper> linkedList = new LinkedList<>();
+
+    final Map<String, ItemWrapper> hashMap = new HashMap<>();
+
+    public DefaultLRUCache() {
+        this.maxSize = DEFAULT_MAX_SIZE;
+        this.timeoutMillis = (long) (DEFAULT_TIMEOUT_SECONDS * 1000);
+    }
+
+    public void setMaxSize(Integer size) {
+        Integer sizeToSet = size;
+        if (size < 0 ) {
+            sizeToSet = DEFAULT_MAX_SIZE;
+        }
+
+        synchronized (lock) {
+            if (linkedList.size() > sizeToSet) {
+                // If cache is bigger than the new maxSize, remove additional items from the end.
+                int itemsToTrim = linkedList.size() - sizeToSet;
+                for (int i = 0; i < itemsToTrim; i++) {
+                    ItemWrapper item = linkedList.removeLast();
+                    hashMap.remove(item.key);
+                }
+            }
+            this.maxSize = sizeToSet;
+        }
+    }
+
+    public void setTimeout(Long timeoutSeconds) {
+        this.timeoutMillis = timeoutSeconds * 1000;
+    }
+
+    public void save(String key, T value) {
+        if (maxSize == 0) {
+            // Cache is disabled when maxSize = 0
+            return;
+        }
+        synchronized (lock) {
+            ItemWrapper item = new ItemWrapper(key, value);
+
+            if (hashMap.containsKey(key)) {
+                ItemWrapper oldItem = hashMap.get(key);
+                linkedList.remove(oldItem);
```

Comment:
Isn't this linear search? It may be a factor in large cache.

Reply:
This is a good catch. Because i was passing the item itself and it uses a doubly linked list, i thought the item should have references to its neighbors and should be able to cleanly remove itself in O(1) but it looks like the actual node with references is managed internally and `remove` actually searched for the item linearly. I then went ahead and redid the whole implementation using `LunkedHashMap`